Footnotes: use core’s meta revisioning if available

adamsilverstein opened the pull request:
<!-- Thanks for contributing to Gutenberg! Please follow the Gutenberg Contributing Guidelines:
https://github.com/WordPress/gutenberg/blob/trunk/CONTRIBUTING.md -->

## What?
<!-- In a few words, what is the PR actually doing? -->
See:
https://core.trac.wordpress.org/ticket/20564 and
https://github.com/WordPress/wordpress-develop/pull/4859

## Why?
<!-- Why is this PR necessary? What problem is it solving? Reference any existing previous issue(s) or PR(s), but please add a short summary here, too -->

## How?
<!-- How is your PR addressing the issue at hand? What are the implementation details? -->

## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard
<!-- How can you test the changes by using the keyboard only? Please note, this is required for PRs that change the user interface (UI). This ensures the PR can be tested for any possible accessibility regressions. -->

## Screenshots or screencast <!-- if applicable -->

adamsilverstein committed:
Footnotes: use core’s meta revisioning if available

adamsilverstein commented:
@ellatrix Since we can't alter this file in core, we need to remove the revisioning here, right? 

I wasn't sure if pattern was to use a WP version check or a capability check like I have here so you can keep the meta revisioning functionality for any older versions of WordPress. What is the correct approach? I can add some inline comments here indicating by the conditional logic is used.

Mamaduka commented:
Hi, Adam

Do you mind releasing this branch on top of the current trunk?

adamsilverstein committed:
Merge branch 'trunk' into redo-footnote-revisioning

# Conflicts:
#	packages/block-library/src/footnotes/index.php

adamsilverstein committed:
Ensure preview menu is closed after opening preview

adamsilverstein commented:
@Mamaduka I rebased onto trunk and also added one line that stabilized the e2e tests. Can you give this another review please?

adamsilverstein committed:
Improve docs and add some new conditionals

adamsilverstein committed:
Add test to verify published post meta is unchanged when previewing meta changes

adamsilverstein commented:
@Mamaduka & @ellatrix 
I added a test in [`c18cae0` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/c18cae0fa6c8d5e72cf05ad15802aa64ddc0e0aa) that fails to illustrate the bug I mentioned on the other ticket: _previewing changes to post meta also writes that meta to the published post_. We should be able to preview meta changes **without** altering a published post.

![image](https://github.com/WordPress/gutenberg/assets/2676022/ea8928eb-6820-419c-b86a-317fc8c86f2b)

adamsilverstein commented:
with the changes in this PR copied over to `src/wp-includes/blocks/footnotes.php`, the issue with overwriting the published post meta goes away so I think there is something in the filters here causing the problem.

adamsilverstein committed:
Correct overwriting of live published post meta when previewing

adamsilverstein commented:
I added a fix for the live published post overwriting issue in [`7d944d2` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/7d944d28fe73b5fe080de9b4ff06cf4a035f0d16)

We can't use `update_post_meta()` to update the autosave's meta, because it doesn't allow revisions; it switches the id to the post parent (which is the live post) and saves the data there :(

You may want to consider back porting this small fix to previous releases (if you do that sort of thing).

adamsilverstein commented:
> @Mamaduka & @ellatrix I added a test in [`c18cae0` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/c18cae0fa6c8d5e72cf05ad15802aa64ddc0e0aa) that fails to illustrate the bug I mentioned on the other ticket: _previewing changes to post meta also writes that meta to the published post_. We should be able to preview meta changes **without** altering a published post.

I added this fix in this standalone PR: https://github.com/WordPress/gutenberg/pull/54339 in case that is helpful

adamsilverstein committed:
Comment out tests that require core changes to pass

adamsilverstein commented:
@Mamaduka I made some small additions to the core code to ensure the autosave changes work correctly. I found some of my testing was invalid because the plugin code was still active in core's footnotes.php file. I think the changes here need to be manually copied over there for testing. 

Can you give this another test, also try commenting back in the last test which I commented out here so this can be merged until the core patch goes in.

Mamaduka commented:
I left the update on the core PR - https://github.com/WordPress/wordpress-develop/pull/4859#issuecomment-1717266655.

adamsilverstein commented:
@Mamaduka / @ellatrix - the core patch has landed in beta1 which should unlock this moving forward - https://core.trac.wordpress.org/changeset/56714

Mamaduka commented:
Fantastic work, @adamsilverstein!

Do you mind rebasing this branch? Now that the code has shipped in the core let's see if anything needs adjustments.

adamsilverstein committed:
Merge branch 'trunk' into redo-footnote-revisioning

adamsilverstein commented:
@Mamaduka do the tests here run against trunk? If so, I should be able to uncomment the section of the 'can be previewed when published' that checks that a published post remains unchanged when previewing meta changes

Mamaduka commented:
@adamsilverstein, yes, tests run against the WP trunk.

adamsilverstein committed:
Include test to verify published post uneffected when previewing meta changes

adamsilverstein commented:
> @adamsilverstein, yes, tests run against the WP trunk.

Great, I enabled the new test section.

Reviewing the GB footnotes PHP now, I see that the only part remaining for WP 6.4+ is the display of footnotes on the revisions screen via the `_wp_post_revision_fields`.

This makes me think we might want to add this display into core as well, ideally in a way that plugins can customize. Right now, the footnotes are simply displayed  directly with `get_metadata( 'post', $revision->ID, $field, true );`. If we automatically displayed revisioned meta on the revisions screen in core, this bit could be removed as well.

Mamaduka commented:
The failing PHPUnit tests look to be revision-related (Global Style Revisions) - https://github.com/WordPress/gutenberg/actions/runs/6315441885/job/17148012985?pr=52988#step:14:63.

adamsilverstein commented:
> The failing PHPUnit tests look to be revision-related (Global Style Revisions) - [WordPress/gutenberg/actions/runs/6315441885/job/17148012985?pr=52988#step:14:63](https://github.com/WordPress/gutenberg/actions/runs/6315441885/job/17148012985?pr=52988#step:14:63).

Hmm, interesting! I'm not familiar with these tests, but it looks like the revisions aren't being created as expected. I'll see if I can get these tests running locally and dig into the underlying cause. I assume these are failing because of the change in trunk not this PR's changes. Please let me know if you see the cause.

adamsilverstein committed:
Global styles tests: fire hooks when updating post

adamsilverstein commented:
>  The failing PHPUnit tests look to be revision-related (Global Style Revisions) - WordPress/gutenberg/actions/runs/6315441885/job/17148012985?pr=52988#step:14:63.

This was a side effect of the way the tests were written. Since revision creation is now hooked on `wp_after_insert_post` revisions are not created when wp_wpdate_post is called with fire_hooks set to false. We will make sure document this change in the dev notes.

I fixed the tests in [`a67bc45` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/a67bc451b2433407e9b4366beb3e3c374a74b000)

Mamaduka reviewed (approved):
Thank you, @adamsilverstein!

I think this is good to merge. Assuming Footnote-related e2e tests aren't failing.

felixarntz commented on packages/block-library/src/footnotes/index.php:
```diff
@@ -9,6 +9,7 @@
  * Renders the `core/footnotes` block on the server.
  *
  * @since 6.3.0
+ * @since 6.4.0 Core added post meta revisions, so this is no longer needed.
```
Is this comment correct here? While it makes sense on the other functions, I think this function continues to be relevant and is also still registered as `render_callback` for the block type.

adamsilverstein replied:
Right, good point - will remove

adamsilverstein replied:
removed

felixarntz reviewed (commented):
@adamsilverstein This looks great to me, just one tiny docs thing.

adamsilverstein commented:
> Assuming Footnote-related e2e tests aren't failing

These are failing now, let me take a look to see if I can tell why.

adamsilverstein committed:
Improve published post navigation in test

adamsilverstein committed:
Comment out failing test for published post

adamsilverstein committed:
Remove incorrect comment line

adamsilverstein commented:
@Mamaduka appreciate your help here. 

I can see the final test failing locally, however when test manually against trunk patched with the update I am proposing here (eg manually editing the footnotes.php file in core to match the code here), repeating the steps of the test the results are as expected. 

Merging the changes here over to core should fix the tests.

Mamaduka commented:
@adamsilverstein, can you highlight which changes?

adamsilverstein commented:
> @adamsilverstein, can you highlight which changes?

The changes that will happen in core when this PR is merged. Basically, copy over the entire contents of Gutenberg `packages/block-library/src/footnotes/index.php`  from this PR to `src/wp-includes/blocks/footnotes.php` in core.

copy that over, rebuild core and rerun the tests, everything should pass.

The reason this is needed is that none of the changes applied in this PR overwrite the footnotes.php.

One possible way to merge this and still have tests work would be to add a block here that removes the callback hooks added in core if they are present and the meta support is present. I'll give that a try.

adamsilverstein committed:
Try: remove filters present in core already

adamsilverstein commented:
In [`cb37c48` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/cb37c4870827f17deb91b8964100cda679022483) I tried removing the filters added by the footnotes code in core, lets see if that helps

Mamaduka commented:
I guess we only need hook removal code until this PR is cherry-picked into the core, correct? In that case, we can put those in a temp file and remove it after the backports. Maybe `compat/plugin/footnotes.php` or something like this.

P.S. I merged #54339 this morning, so you might want to rebase the branch.

adamsilverstein commented:
> I guess we only need hook removal code until this PR is cherry-picked into the core, correct? In that case, we can put those in a temp file and remove it after the backports. Maybe compat/plugin/footnotes.php or something like this.

Right, good suggestion. Let me see if this fixes the tests first then I will try that.

adamsilverstein committed:
Merge branch 'trunk' into redo-footnote-revisioning

# Conflicts:
#	packages/block-library/src/footnotes/index.php

adamsilverstein committed:
Move footnote filter shim to an external compat file

adamsilverstein committed:
phpcbf

adamsilverstein committed:
Remove remaining hooks from footnotes that are causing issues

adamsilverstein commented:
@Mamaduka - I finally figured out what is happening here by running the tests locally - autosaving is not working correctly when permalinks are not enabled. I paused the in-flight tests, enabled permalinks and the tests passed. Pretty sure this is an underlying bug the tests have caught I will try to address. In the meantime, I'm also going to try running the tests with permalinks enabled to verify that is the cause of failure.

adamsilverstein committed:
Revert "Remove remaining hooks from footnotes that are causing issues"

This reverts commit 9d2f1200f8b9030581e89904dfcf084dfe251d29.

Mamaduka commented:
@adamsilverstein, that's some fine detective work here!

adamsilverstein commented:
> @adamsilverstein, that's some fine detective work here!

We'll see if I am right

adamsilverstein commented:
> > @adamsilverstein, that's some fine detective work here!
> 
> We'll see if I am right

Not sure this is right, testing in core everything works fine without permalinks enabled. I do have tests failing locally so at least I have something to work on/test against.

I'll keep working on this and let you know what I find, appreciate any help!

adamsilverstein committed:
Temporarily disable slashing test

adamsilverstein committed:
Restore test to verify that the published post is unchanged after previewing changes to meta.

adamsilverstein committed:
Move the meta revision compat code to the footnotes shim file

This preveents the code from being included in core.

This code can be removed once GB no longer supports core < 6.4.

adamsilverstein commented:
@Mamaduka turns out the issue was slashing! 

Rerunning tests without the slash test to confirm - temporarily in https://github.com/WordPress/gutenberg/pull/52988/commits/7fee4f57a752fa4af44eda471a2027b927196106

This will fix the issue: https://github.com/WordPress/wordpress-develop/pull/5344

Also, I realized we should move all of the revision related functionality to the shim so we don't bring that over to core. I did that in [`9d2f120` (#52988)](https://github.com/WordPress/gutenberg/pull/52988/commits/9d2f1200f8b9030581e89904dfcf084dfe251d29)

adamsilverstein committed:
remove hooks from footnotes file

adamsilverstein committed:
Ensure functions are guarded against redeclaration.

adamsilverstein committed:
phpcbf

adamsilverstein committed:
phpcbf2

adamsilverstein committed:
restore typing “3”

adamsilverstein committed:
remove extra space

adamsilverstein commented:
@Mamaduka E2E tests now pass (without the slash test). Once I merge the fix to core in https://github.com/WordPress/wordpress-develop/pull/5344 we can add the slash test back. If you have a chance can you please review that PR? I added a test to validate the slashing bug as well as manual reproduction instructions.

~I see some unit tests failures but it looks like flakey tests so I'll try rerunning.~

All tests are now green :)

Mamaduka reviewed (approved):
Thank you, @adamsilverstein! 🦸

I assume commented-out tests can be enabled once the core PR is committed.

The PHP shims might need re-arrangement so Footnote revisions could work with WP < 6.4. I can do that after this is cherry-picked in beta two 👍

P.S. We can ignore the optional CI check; I am unsure why it's failing.

adamsilverstein commented:
> I assume commented-out tests can be enabled once the core PR is committed.

Yes, exactly.

adamsilverstein commented:
> I assume commented-out tests can be enabled once the core PR is committed.

Committed in https://core.trac.wordpress.org/changeset/56745 so all the tests should pass now, I will uncomment the slash tests.

adamsilverstein committed:
Restore slashed data test

adamsilverstein commented:
> The PHP shims might need re-arrangement so Footnote revisions could work with WP < 6.4. I can do that after this is cherry-picked in beta two 👍

Feel free to rework, I wasn't really sure what the best approach was and I didn't test in WP < 6.4. I do like the idea of a shim file that isn't ported over to core. I'd like to add the last bit of functionality that displays the meta on the revisions screen in core as well, so it won't need to be hooked in by the plugin. I'll work on that soon.

adamsilverstein commented:
@Mamaduka - the slash test is restored and all E2E tests are green, this should be ready to go - shall I go ahead and click "Squash and Merge"?

Mamaduka commented:
@adamsilverstein, sounds good 👍

mikachan commented:
I just cherry-picked this PR to the commits-for-6.4-beta2 branch to get it included in the next release: 507d0e9b779d0a95b228ec23bf54f5ead36e8e6f

ellatrix commented:
Great work @adamsilverstein, thank you!

youknowriad commented on lib/compat/plugin/footnotes.php:
```diff
@@ -0,0 +1,250 @@
+<?php
```
@ellatrix If I'm not wrong this file should be moved to `lib/compat/wordpress-6.4` because it's just here to support versions priori to 6.4 right?

ellatrix replied:
Yes, although maybe I don't think these make sense anymore after the revision support was added. Cc @adamsilverstein
